In [1]:
print(1)

1


In [14]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt


pd.set_option('display.max_columns', None)

In [17]:
df = pd.read_excel(r"F:\Learn\cgpa_analysis\Data.xlsx")
df.head()

Timestamp                 Branch     Year?  \
0 2025-05-04 21:48:31                     IT  4th year   
1 2025-05-04 21:50:20      CSE (and its spl)  4th year   
2 2025-05-04 21:51:01      ECE (and its spl)  4th year   
3 2025-05-04 21:51:21      ECE (and its spl)  4th year   
4 2025-05-04 21:52:27  EEE/EIE (and its spl)  4th year   

  Does your degree have a specialization?  \
0                                    Nope   
1                                     Yep   
2                                     Yep   
3                                     Yep   
4                                    Nope   

                                     1. Class Notes?  \
0  One notebook for all subjects. "If I feel like...   
1                  What notes? [no notes maintained]   
2         I carry 3 pens to class. [Pro level notes]   
3  I make notes that I can understand. [Well main...   
4  One notebook for all subjects. "If I feel like...   

                                              2. DA? 3. CGPA  \
0  5+ instances where submission was done one day...    7.85   
1      Login: 11:55 => Submit: 11:59 [Saved in time]    9.15   
2        I panic if it's not done by 11pm. [On time]     9.1   
3        I panic if it's not done by 11pm. [On time]    8.65   
4      Login: 11:55 => Submit: 11:59 [Saved in time]    7.38   

   4. Number of Backlogs?  5. Average Attendance? (%)  \
0                     5.0                           7   
1                     0.0                           3   
2                     1.0                           9   
3                     2.0                           9   
4                     0.0                           8   

                                        6. FFCS prep  \
0  I choose slots on the day of FFCS and hope the...   
1  I check profs, slots and have backup timetable...   
2  I check profs, slots and have backup timetable...   
3  I check profs, slots and have backup timetable...   
4  I check profs, slots and have backup timetable...   

                                        7. Exam Prep  \
0  I download the syllabus one day before. [Last ...   
1  I download the syllabus one day before. [Last ...   
2  I start anytime in the previous week of exam. ...   
3  I start anytime in the previous week of exam. ...   
4  I download the syllabus one day before. [Last ...   

  8. Active member of Clubs/Chapters/Teams/Anything that involved Night-slips or ODs( events etc)?  \
0                                Yes, I just want OD                                                 
1    Yes, but I compensate for my academics somehow.                                                 
2    Yes, but I compensate for my academics somehow.                                                 
3                    Yes, I want to gain experience.                                                 
4                    Yes, I want to gain experience.                                                 

   9. Number of events participated in like Hackathons/Ideathons / Events related to my degree or career?  \
0                                                2.0                                                        
1                                                2.0                                                        
2                                                2.0                                                        
3                                               10.0                                                        
4                                               10.0                                                        

               10. What's the perfect seat for your?  \
0                                 First/second bench   
1  If there is something beyond last bench, I'd c...   
2                                 First/second bench   
3                                 First/second bench   
4                            Somewhere in the middle   

                       11. Bond with teachers.  

In [18]:
df.columns

Index(['Timestamp', 'Branch', 'Year?',
       'Does your degree have a specialization?', '1. Class Notes?', '2. DA?',
       '3. CGPA', '4. Number of Backlogs?', '5. Average Attendance? (%)',
       '6. FFCS prep', '7. Exam Prep',
       '8. Active member of Clubs/Chapters/Teams/Anything that involved Night-slips or ODs( events etc)?',
       '9. Number of events participated in like Hackathons/Ideathons / Events related to my degree or career?',
       '10. What's the perfect seat for your?', '11. Bond with teachers.',
       '12. Study sources?', '13. Any disciplinary action?',
       '14. Hanging out with friends after classes?',
       '15. Sleep and Exercise?', '16. Study location?',
       '17. Room type( dayscholar/ single bed/4bed/6bed etc)'],
      dtype='object')

In [19]:
df.rename(columns={
    'Year?':'year',
     'Does your degree have a specialization?':"spl",
     '1. Class Notes?':"class_notes",
     '2. DA?':"da",
     '3. CGPA':"cgpa",
     '4. Number of Backlogs?':'backlogs',
     '5. Average Attendance? (%)':"attendance",
     '6. FFCS prep':"ffcs",
      '7. Exam Prep':"exam_prep",
      '8. Active member of Clubs/Chapters/Teams/Anything that involved Night-slips or ODs( events etc)?':"clubs_chapters",
      '9. Number of events participated in like Hackathons/Ideathons / Events related to my degree or career?':"competitions",
      "10. What's the perfect seat for your?":"seating_arrangement",
      '11. Bond with teachers.':"bond_teachers",
      '12. Study sources?':"study_material",
      '13. Any disciplinary action?':"disciplinary_action",
       '14. Hanging out with friends after classes?':"social_life",
       '15. Sleep and Exercise?':"lifestyle",
        '16. Study location?': "study_location",
        '17. Room type( dayscholar/ single bed/4bed/6bed etc)':"room_type"

}, inplace=True)

In [20]:
df.head()

Timestamp                 Branch      year   spl  \
0 2025-05-04 21:48:31                     IT  4th year  Nope   
1 2025-05-04 21:50:20      CSE (and its spl)  4th year   Yep   
2 2025-05-04 21:51:01      ECE (and its spl)  4th year   Yep   
3 2025-05-04 21:51:21      ECE (and its spl)  4th year   Yep   
4 2025-05-04 21:52:27  EEE/EIE (and its spl)  4th year  Nope   

                                         class_notes  \
0  One notebook for all subjects. "If I feel like...   
1                  What notes? [no notes maintained]   
2         I carry 3 pens to class. [Pro level notes]   
3  I make notes that I can understand. [Well main...   
4  One notebook for all subjects. "If I feel like...   

                                                  da  cgpa  backlogs  \
0  5+ instances where submission was done one day...  7.85       5.0   
1      Login: 11:55 => Submit: 11:59 [Saved in time]  9.15       0.0   
2        I panic if it's not done by 11pm. [On time]   9.1       1.0   
3        I panic if it's not done by 11pm. [On time]  8.65       2.0   
4      Login: 11:55 => Submit: 11:59 [Saved in time]  7.38       0.0   

   attendance                                               ffcs  \
0           7  I choose slots on the day of FFCS and hope the...   
1           3  I check profs, slots and have backup timetable...   
2           9  I check profs, slots and have backup timetable...   
3           9  I check profs, slots and have backup timetable...   
4           8  I check profs, slots and have backup timetable...   

                                           exam_prep  \
0  I download the syllabus one day before. [Last ...   
1  I download the syllabus one day before. [Last ...   
2  I start anytime in the previous week of exam. ...   
3  I start anytime in the previous week of exam. ...   
4  I download the syllabus one day before. [Last ...   

                                    clubs_chapters  competitions  \
0                              Yes, I just want OD           2.0   
1  Yes, but I compensate for my academics somehow.           2.0   
2  Yes, but I compensate for my academics somehow.           2.0   
3                  Yes, I want to gain experience.          10.0   
4                  Yes, I want to gain experience.          10.0   

                                 seating_arrangement  \
0                                 First/second bench   
1  If there is something beyond last bench, I'd c...   
2                                 First/second bench   
3                                 First/second bench   
4                            Somewhere in the middle   

                                 bond_teachers  \
0                 They usually know me by face   
1           They wouldn't recognize me at all.   
2  They'd notice if I didn't attend the class.   
3                 They usually know me by face   
4  They'd notice if I didn't attend the class.   

                           study_material disciplinary_action  \
0  Important topic slides, youtube, books                 Yes   
1                  All slides and youtube                  No   
2                  All slides and youtube                  No   
3  Important topic slides, youtube, books                  No   
4  Important topic slides, youtube, books                  No   

                                         social_life  \
0  I exhaust my outing time and I wish there was ...   
1                            30min - 1hr on somedays   
2                            30min - 1hr on somedays   
3                            30min - 1hr on somedays   
4  I exhaust my outing time and I wish there was ...   

                                 lifestyle study_location  room_type  
0            8+ hours | Exercise regularly           Room        6.0  
1        4-6 hours | Occasionally exercise           Room        2.0  
2        4-6 hours | Occasionally exercise           Room        4.0  
3  6-8 hours | Exercise a few times a week        Li

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Timestamp            200 non-null    datetime64[ns]
 1   Branch               200 non-null    object        
 2   year                 200 non-null    object        
 3   spl                  200 non-null    object        
 4   class_notes          199 non-null    object        
 5   da                   200 non-null    object        
 6   cgpa                 200 non-null    object        
 7   backlogs             199 non-null    float64       
 8   attendance           200 non-null    int64         
 9   ffcs                 199 non-null    object        
 10  exam_prep            198 non-null    object        
 11  clubs_chapters       198 non-null    object        
 12  competitions         197 non-null    float64       
 13  seating_arrangement  198 non-null  

Nulls at:
class notes, backlogs, ffcs, exam_prep,club_chapters,competitions, seating arrangement, bong_teachers, study_material, disciplinary_action,
social_life, lifestyle, study_location, room_type

cgpa is object. Should have been float. 
this field is an indicator to bogus data points. rows with incorrect entries can be dropped. 

In [21]:
df['cgpa'][df['cgpa'].apply(lambda x: isinstance(x,str))]

6         Kya karoge jaan ke?
22                     8.36.0
61                          😛
75              Who kya howe?
83     Nahi bataunga(jk 8.52)
95                   *secret*
115    Nah dude too personal.
122                     8-8.5
132             Dont remember
148                        9+
158                  Nice try
Name: cgpa, dtype: object

We need to handle the information given to us. 
- Some of them are genuine. Eg 8.3.3 (typo)
- 9+ (will take 9)
- Nahi bataunga(jk 8.52)	
- 8-8.5	

regex maybe?

In [63]:
import re
def clean_gpa(gpa):
    gpa = gpa.strip()
    s =  re.search(r"\d| \d.\d|\d.\d\d", gpa)
    # i = float(.join(s).strip())
    return s
clean_gpa("9+ Who kya howe? 8.9")


<re.Match object; span=(0, 1), match='9'>

In [44]:
df.head()

Timestamp                 Branch      year   spl  \
0 2025-05-04 21:48:31                     IT  4th year  Nope   
1 2025-05-04 21:50:20      CSE (and its spl)  4th year   Yep   
2 2025-05-04 21:51:01      ECE (and its spl)  4th year   Yep   
3 2025-05-04 21:51:21      ECE (and its spl)  4th year   Yep   
4 2025-05-04 21:52:27  EEE/EIE (and its spl)  4th year  Nope   

                                         class_notes  \
0  One notebook for all subjects. "If I feel like...   
1                  What notes? [no notes maintained]   
2         I carry 3 pens to class. [Pro level notes]   
3  I make notes that I can understand. [Well main...   
4  One notebook for all subjects. "If I feel like...   

                                                  da  cgpa  backlogs  \
0  5+ instances where submission was done one day...  7.85       5.0   
1      Login: 11:55 => Submit: 11:59 [Saved in time]  9.15       0.0   
2        I panic if it's not done by 11pm. [On time]   9.1       1.0   
3        I panic if it's not done by 11pm. [On time]  8.65       2.0   
4      Login: 11:55 => Submit: 11:59 [Saved in time]  7.38       0.0   

   attendance                                               ffcs  \
0           7  I choose slots on the day of FFCS and hope the...   
1           3  I check profs, slots and have backup timetable...   
2           9  I check profs, slots and have backup timetable...   
3           9  I check profs, slots and have backup timetable...   
4           8  I check profs, slots and have backup timetable...   

                                           exam_prep  \
0  I download the syllabus one day before. [Last ...   
1  I download the syllabus one day before. [Last ...   
2  I start anytime in the previous week of exam. ...   
3  I start anytime in the previous week of exam. ...   
4  I download the syllabus one day before. [Last ...   

                                    clubs_chapters  competitions  \
0                              Yes, I just want OD           2.0   
1  Yes, but I compensate for my academics somehow.           2.0   
2  Yes, but I compensate for my academics somehow.           2.0   
3                  Yes, I want to gain experience.          10.0   
4                  Yes, I want to gain experience.          10.0   

                                 seating_arrangement  \
0                                 First/second bench   
1  If there is something beyond last bench, I'd c...   
2                                 First/second bench   
3                                 First/second bench   
4                            Somewhere in the middle   

                                 bond_teachers  \
0                 They usually know me by face   
1           They wouldn't recognize me at all.   
2  They'd notice if I didn't attend the class.   
3                 They usually know me by face   
4  They'd notice if I didn't attend the class.   

                           study_material disciplinary_action  \
0  Important topic slides, youtube, books                 Yes   
1                  All slides and youtube                  No   
2                  All slides and youtube                  No   
3  Important topic slides, youtube, books                  No   
4  Important topic slides, youtube, books                  No   

                                         social_life  \
0  I exhaust my outing time and I wish there was ...   
1                            30min - 1hr on somedays   
2                            30min - 1hr on somedays   
3                            30min - 1hr on somedays   
4  I exhaust my outing time and I wish there was ...   

                                 lifestyle study_location  room_type  
0            8+ hours | Exercise regularly           Room        6.0  
1        4-6 hours | Occasionally exercise           Room        2.0  
2        4-6 hours | Occasionally exercise           Room        4.0  
3  6-8 hours | Exercise a few times a week        Li

In [45]:
x = 1
(isinstance(x,float) or isinstance(x,int))

True

In [46]:
temp = df['cgpa'].apply(lambda x: x if (isinstance(x,float) or isinstance(x,int)) else clean_gpa(x))[df['cgpa'].apply(lambda x: isinstance(x,str))]
temp


IndexError: list index out of range